File: chat/src/main/service/saga_ack_waiter.py
"""
SAGA ACK waiter for Python → Kotlin cross-DB operations.

When Python initiates a cross-DB operation (e.g., model provider CRUD),
it publishes to a Redis Stream and waits for Kotlin to ACK via the
saga_ack stream before committing locally.
"""

import time

from src.main.utils.core.logger import get_logger

logger = get_logger(__name__)

_SAGA_ACK_STREAM = "scrapalot:stream:saga_ack"
_CONSUMER_GROUP = "cg-scrapalot-chat-saga"
_CONSUMER_NAME = "scrapalot-chat-saga-0"


def _ensure_consumer_group(redis_client) -> None:
    """Idempotent creation of the saga_ack consumer group for Python."""
    import redis as redis_lib

    try:
        redis_client.xgroup_create(_SAGA_ACK_STREAM, _CONSUMER_GROUP, id="0", mkstream=True)
    except redis_lib.ResponseError as e:
        if "BUSYGROUP" not in str(e):
            raise


def wait_for_saga_ack(saga_id: str, timeout: float = 10.0) -> dict | None:
    """
    Block until a SAGA ACK with matching saga_id appears in the stream.

    Args:
        saga_id: The SAGA ID to wait for.
        timeout: Max seconds to wait.

    Returns:
        Dict with 'status' ('ACK' or 'NACK') and optional 'error', or None on timeout.
    """
    from src.main.utils.redis.client import get_redis_client

    redis_client = get_redis_client()
    _ensure_consumer_group(redis_client)

    deadline = time.monotonic() + timeout

    while time.monotonic() < deadline:
        remaining = deadline - time.monotonic()
        block_ms = int(min(remaining * 1000, 500))
        if block_ms <= 0:
            break

        try:
            results = redis_client.xreadgroup(
                groupname=_CONSUMER_GROUP,
                consumername=_CONSUMER_NAME,
                streams={_SAGA_ACK_STREAM: ">"},
                count=10,
                block=block_ms,
            )

            if not results:
                continue

            # noinspection PyTypeChecker
            for _stream_key_raw, messages in results:
                for message_id, fields in messages:
                    # Decode bytes
                    decoded = {}
                    for k, v in fields.items():
                        key = k.decode("utf-8") if isinstance(k, bytes) else k
                        val = v.decode("utf-8") if isinstance(v, bytes) else v
                        decoded[key] = val

                    # Always ACK — even if not our saga_id
                    redis_client.xack(_SAGA_ACK_STREAM, _CONSUMER_GROUP, message_id)

                    if decoded.get("saga_id") == saga_id:
                        status = decoded.get("status", "NACK")
                        error = decoded.get("error")
                        logger.debug("Received SAGA %s for saga_id=%s", status, saga_id)
                        return {"status": status, "error": error}

        except Exception as e:
            logger.warning("Error reading saga_ack stream: %s", e)
            time.sleep(0.1)

    logger.warning("SAGA ACK timeout for saga_id=%s after %.1fs", saga_id, timeout)
    return None

File: backend/src/main/kotlin/com/scrapalot/backend/service/SagaAckWaiter.kt
package com.scrapalot.backend.service

import com.scrapalot.backend.utils.ensureConsumerGroup
import jakarta.annotation.PostConstruct
import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.data.redis.connection.stream.Consumer
import org.springframework.data.redis.connection.stream.ReadOffset
import org.springframework.data.redis.connection.stream.StreamOffset
import org.springframework.data.redis.connection.stream.StreamReadOptions
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.stereotype.Component
import java.time.Duration
import java.time.Instant

private val logger = KotlinLogging.logger {}

data class SagaAck(
    val status: String,
    val error: String? = null
)

/**
 * Waits for SAGA acknowledgements from the saga_ack stream.
 *
 * Used by Kotlin services to block until Python confirms
 * it has processed and committed a cross-DB operation.
 */
@Component
class SagaAckWaiter(
    @param:Qualifier("streamStringRedisTemplate")
    private val stringRedisTemplate: StringRedisTemplate
) {
    companion object {
        const val SAGA_ACK_STREAM = RedisEventPublisher.STREAM_SAGA_ACK
        const val CONSUMER_GROUP = "cg-scrapalot-backend-saga"
        const val CONSUMER_NAME = "scrapalot-backend-saga-0"
    }

    @PostConstruct
    fun init() {
        ensureConsumerGroup(stringRedisTemplate, SAGA_ACK_STREAM, CONSUMER_GROUP)
    }

    /**
     * Block until an ACK with matching saga_id appears in the stream,
     * or until timeout expires.
     *
     * Returns null on timeout.
     */
    fun waitForAck(
        sagaId: String,
        timeout: Duration = Duration.ofSeconds(10)
    ): SagaAck? {
        val deadline = Instant.now().plus(timeout)
        val ops = stringRedisTemplate.opsForStream<String, String>()

        while (Instant.now().isBefore(deadline)) {
            val remaining = Duration.between(Instant.now(), deadline)
            val blockMs = remaining.toMillis().coerceAtMost(500).coerceAtLeast(50)

            try {
                val results =
                    ops.read(
                        Consumer.from(CONSUMER_GROUP, CONSUMER_NAME),
                        StreamReadOptions.empty().count(10).block(Duration.ofMillis(blockMs)),
                        StreamOffset.create(SAGA_ACK_STREAM, ReadOffset.lastConsumed())
                    )

                for (record in results ?: emptyList()) {
                    // Always acknowledge — even if it's for a different saga, we've consumed it
                    ops.acknowledge(CONSUMER_GROUP, record)

                    val recordSagaId = record.value["saga_id"]
                    if (recordSagaId == sagaId) {
                        val status = record.value["status"] ?: "NACK"
                        val error = record.value["error"]
                        logger.debug { "Received SAGA ACK for $sagaId: status=$status" }
                        return SagaAck(status, error)
                    }
                }
            } catch (e: Exception) {
                logger.warn { "Error reading saga_ack stream: ${e.message}" }
                Thread.sleep(100)
            }
        }

        logger.warn { "SAGA ACK timeout for saga_id=$sagaId after ${timeout.seconds}s" }
        return null
    }
}
